csrc/fusion_profiler.cpp

```diff
@@ -0,0 +1,787 @@
+           << " " << std::setw(9) << "-"
+           << " " << std::setw(9) << "-"
+           << " " << std::setw(11) << "-"
+           << " " << std::setw(9) << "-";
+      }
+      std::stringstream grid;
+      grid << "[" << std::get<0>(kp.grid) << ", " << std::get<1>(kp.grid)
+           << ", " << std::get<2>(kp.grid) << "]";
+      std::stringstream block;
+      block << "[" << std::get<0>(kp.block) << ", " << std::get<1>(kp.block)
+            << ", " << std::get<2>(kp.block) << "]";
+      std::stringstream cluster;
+      cluster << "[" << std::get<0>(kp.cluster) << ", "
+              << std::get<1>(kp.cluster) << ", " << std::get<2>(kp.cluster)
+              << "]";
+      std::stringstream smem;
+      smem << "[" << kp.dynamic_shared_mem << ", " << kp.static_shared_mem
+           << "]";
+      os << std::setfill(' ') << std::right << std::fixed << " " << std::setw(6)
+         << idx << " " << std::setw(10) << kp.name << " " << std::setw(11)
+         << std::setprecision(3) << kp.time_ms;
+
+      if (fp.verbose) {
+        os << " " << std::setw(11) << std::setprecision(3)
+           << kp.compile_time_ms;
+      }
+
+      os << " " << std::setw(13) << std::setprecision(2)
+         << kp.effective_bandwidth_gbs << " " << std::setw(9)
+         << std::setprecision(2) << kp.percentage_peak_bandwidth << " "
+         << std::setw(9) << std::setprecision(3)
+         << ((double)kp.input_bytes / 1000000.0) << " " << std::setw(9)
+         << std::setprecision(3) << ((double)kp.output_bytes / 1000000.0) << " "
+         << std::setw(16) << smem.str() << " " << std::setw(6) << kp.registers
+         << " " << std::setw(16) << grid.str() << " " << std::setw(16)
+         << block.str();
+      if (fp.verbose) {
+        os << " " << std::setw(16) << cluster.str() << " " << std::setw(5)
+           << kp.device << " " << std::setw(5) << kp.stream << " "
+           << std::setw(14) << std::setprecision(2) << kp.peak_bandwidth_gbs
+           << " " << std::setw(20) << kp.device_name;
+      }
+      os << std::endl;
+      ++idx;
+    }
+  }
+  return os;
+}
+
+FusionProfiler* FusionProfiler::singleton_ = nullptr;
+
+FusionProfiler::FusionProfiler()
+    : cupti_disabled_(false),
+      cupti_buffer_(FusionProfiler::cupti_activity_buffer_size),
+      state_(ProfilerState::Ready),
+      fusion_id_(-1),
+      profile_(),
+      fusion_timer_(at::cuda::getCurrentCUDAStream()),
+      host_timer_(),
+      compile_timer_(),
+      segments_(),
+      device_descriptors_(),
+      kernel_profiles_(),
+      corrid_2_segid_() {
+  if (!cupti_disabled_) {
+    NVFUSER_CUPTI_SAFE_CALL(cuptiActivityRegisterCallbacks(
+        cupti_buffer_requested, cupti_buffer_completed));
+  }
+}
+
+FusionProfiler* FusionProfiler::get() {
+  static std::mutex singleton_lock;
+  std::lock_guard<std::mutex> guard(singleton_lock);
+  if (singleton_ == nullptr) {
+    singleton_ = new FusionProfiler();
+  }
+  return singleton_;
+}
```

Comment:
It appears that `singleton_` is only used in this function. I'd just define it here, getting rid of a non-local variable. 

```suggestion
  static std::mutex singleton_lock;
  static FusionProfiler* singleton = nullptr;
  
  std::lock_guard<std::mutex> guard(singleton_lock);
  if (singleton == nullptr) {
    singleton = new FusionProfiler();
  }
  return singleton;
}
```

Reply:
Change made!